types/evidence.go

```diff
@@ -1167,23 +1217,36 @@ func (e ProofOfLockChange) BlockID() BlockID {
 	return e.Votes[0].BlockID
 }
 
-// In order for a ProofOfLockChange to be valid, a validator must have received +2/3 majority of votes
-// MajorityOfVotes checks that there were sufficient votes in order to change locks
-func (e ProofOfLockChange) MajorityOfVotes(valSet *ValidatorSet) bool {
+// ValidateVotes checks the polc against the validator set of that height. The function makes sure that the polc
+// contains a majority of votes and that each
+func (e ProofOfLockChange) ValidateVotes(valSet *ValidatorSet, chainID string) error {
+	if e.IsAbsent() {
+		return errors.New("polc is empty")
+	}
 	talliedVotingPower := int64(0)
 	votingPowerNeeded := valSet.TotalVotingPower() * 2 / 3
-	for _, validator := range valSet.Validators {
-		for _, vote := range e.Votes {
+	for _, vote := range e.Votes {
+		exists := false
+		for _, validator := range valSet.Validators {
 			if bytes.Equal(validator.Address, vote.ValidatorAddress) {
-				talliedVotingPower += validator.VotingPower
-
-				if talliedVotingPower > votingPowerNeeded {
-					return true
+				exists = true
+				if !validator.PubKey.VerifyBytes(vote.SignBytes(chainID), vote.Signature) {
+					return fmt.Errorf("cannot verify vote (from validator: %d) against signature: %v",
+						vote.ValidatorIndex, vote.Signature)
 				}
+
+				talliedVotingPower += validator.VotingPower
 			}
 		}
+		if !exists {
+			return fmt.Errorf("vote was not from a validator in this set: %v", vote.String())
```

Comment:
can we print validator as well?

Reply:
Sure

Reply:
Do you mean the validator address of the vote?

Reply:
forgot we have it in vote, ignore